Code/Tools/Libs/ToolsFoundation/FileSystem/Implementation/FileSystemModel.cpp

```diff
@@ -0,0 +1,855 @@
+    }
+
+    return {};
+  }
+
+  return HandleSingleFile(sAbsolutePath, Stats, bRecurseIntoFolders);
+}
+
+
+ezFileStatus ezFileSystemModel::HandleSingleFile(const ezString& sAbsolutePath, const ezFileStats& FileStat, bool bRecurseIntoFolders)
+{
+  FILESYSTEM_PROFILE("HandleSingleFile2");
+
+  if (FileStat.m_bIsDirectory)
+  {
+    ezFileStatus status;
+    bool bExisted = false;
+    {
+      EZ_LOCK(m_FilesMutex);
+      auto it = m_ReferencedFolders.FindOrAdd(sAbsolutePath, &bExisted);
+      it.Value() = ezFileStatus::Status::Valid;
+    }
+
+    if (!bExisted)
+    {
+      FireFolderChangedEvent(sAbsolutePath, ezFolderChangedEvent::Type::FolderAdded);
+      if (bRecurseIntoFolders)
+        CheckFolder(sAbsolutePath);
+    }
+
+    return {};
+  }
+  else
+  {
+    ezFileStatus status;
+    bool bExisted = false;
+    bool bFileChanged = false;
+    {
+      EZ_LOCK(m_FilesMutex);
+      auto it = m_ReferencedFiles.FindOrAdd(sAbsolutePath, &bExisted);
+      ezFileStatus& value = it.Value();
+      bFileChanged = !value.m_Timestamp.Compare(FileStat.m_LastModificationTime, ezTimestamp::CompareMode::Identical);
+      if (bFileChanged)
+      {
+        value.m_uiHash = 0;
+      }
+
+      // If the state is unknown, we loaded it from the cache and need to fire FileChanged to update dependent systems.
+      // #TODO_ASSET This behaviors should be changed once the asset cache is stored less lossy.
+      bFileChanged |= value.m_Status == ezFileStatus::Status::Unknown;
+      // mark the file as valid (i.e. we saw it on disk, so it hasn't been deleted or such)
+      value.m_Status = ezFileStatus::Status::Valid;
+      value.m_Timestamp = FileStat.m_LastModificationTime;
+      status = value;
+    }
+
+    if (!bExisted)
+    {
+      FireFileChangedEvent(sAbsolutePath, status, ezFileChangedEvent::Type::FileAdded);
+    }
+    else if (bFileChanged)
+    {
+      FireFileChangedEvent(sAbsolutePath, status, ezFileChangedEvent::Type::FileChanged);
+    }
+    return status;
+  }
+}
+
+void ezFileSystemModel::MarkFileLocked(ezStringView sAbsolutePath)
+{
+  EZ_LOCK(m_FilesMutex);
+  auto it = m_ReferencedFiles.Find(sAbsolutePath);
+  if (it.IsValid())
+  {
+    it.Value().m_Status = ezFileStatus::Status::FileLocked;
+    m_LockedFiles.Insert(sAbsolutePath);
+  }
+}
+
+void ezFileSystemModel::FireFileChangedEvent(ezStringView sFile, ezFileStatus fileStatus, ezFileChangedEvent::Type type)
+{
+  // We queue up all requests on a thread and only return once the list is empty. The reason for this is that:
+  // A: We don't want to allow recursive event calling as it creates limbo states in the model and hard to debug bugs.
+  // B: If a user calls NotifyOfChange, the function should only return if the event and any indirect events that were triggered by the event handlers have been processed.
+  static thread_local ezHybridArray<ezFileChangedEvent, 2, ezStaticAllocatorWrapper> m_PostponedFiles;
```

Comment:
static thread_local might not be allowed on all compilers. Move the variable out of the function and just make it a thread_local global variable.